File: plugins/review-fixer/src/resolve.ts
// resolve.ts — discover token / repo / PR from flags first, then environment,
// then the gh/git CLIs. Each resolver takes the Deps seam so tests can stub the
// world.
import type { Deps } from "./deps"

/** Resolve a GitHub token: flag → GH_TOKEN → GITHUB_TOKEN → `gh auth token`. */
export async function resolveToken(deps: Deps, flagToken?: string): Promise<string> {
  if (flagToken && flagToken.trim() !== "") return flagToken
  if (deps.env.GH_TOKEN && deps.env.GH_TOKEN.trim() !== "") return deps.env.GH_TOKEN
  if (deps.env.GITHUB_TOKEN && deps.env.GITHUB_TOKEN.trim() !== "") return deps.env.GITHUB_TOKEN
  const { stdout, exitCode } = await deps.runCmd(["gh", "auth", "token"])
  if (exitCode !== 0) throw new Error("resolve token: gh auth token failed")
  const tok = stdout.trim()
  if (tok === "") throw new Error("no GitHub token available (set GH_TOKEN, GITHUB_TOKEN, or run `gh auth login`)")
  return tok
}

/** Resolve the repo as owner/name: flag → `gh repo view`. */
export async function resolveRepo(
  deps: Deps,
  flagRepo?: string,
): Promise<{ owner: string; name: string }> {
  let spec: string
  if (flagRepo && flagRepo.trim() !== "") {
    spec = flagRepo.trim()
  } else {
    const { stdout, exitCode } = await deps.runCmd(["gh", "repo", "view", "--json", "nameWithOwner", "-q", ".nameWithOwner"])
    if (exitCode !== 0) throw new Error("resolve repo: gh repo view failed")
    spec = stdout.trim()
  }
  const slash = spec.indexOf("/")
  // Both empty-owner ("/name") and empty-name ("owner/") collapse into the
  // same check: slash must be strictly between the ends.
  if (slash <= 0 || slash === spec.length - 1) {
    throw new Error(`invalid repo "${spec}": want owner/name`)
  }
  return { owner: spec.slice(0, slash), name: spec.slice(slash + 1) }
}

/** Resolve a PR number: flag → branch → `gh pr list --head <branch>`. */
export async function resolvePR(deps: Deps, cwd: string, flagPR?: number): Promise<number> {
  if (flagPR !== undefined && flagPR !== 0) return flagPR
  const { stdout: branchOut, exitCode: branchExit } = await deps.runCmd(["git", "branch", "--show-current"], { cwd })
  if (branchExit !== 0) throw new Error("resolve pr: not on a branch (detached HEAD?); pass pr")
  const branch = branchOut.trim()
  if (branch === "") throw new Error("resolve pr: not on a branch (detached HEAD?); pass pr")
  const { stdout: prOut, exitCode: prExit } = await deps.runCmd(
    ["gh", "pr", "list", "--head", branch, "--json", "number", "-q", ".[0].number"],
    { cwd },
  )
  if (prExit !== 0) throw new Error(`resolve pr: no open PR found for branch "${branch}"`)
  const s = prOut.trim()
  if (s === "") throw new Error(`resolve pr: no open PR found for branch "${branch}"`)
  const n = Number.parseInt(s, 10)
  if (Number.isNaN(n)) throw new Error(`resolve pr: parse "${s}" failed`)
  return n
}

File: plugins/review-fixer/reviewfixer/resolve.go
package reviewfixer

import (
	"context"
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

// execTimeout bounds how long an exec'd git/gh helper may run before it is cancelled.
const execTimeout = 30 * time.Second

// runCmd runs an external command through r under a fresh timeout so a hung git/gh
// invocation cannot block the process forever.
func runCmd(ctx context.Context, r Runner, name string, args ...string) (string, error) {
	cctx, cancel := context.WithTimeout(ctx, execTimeout)
	defer cancel()

	return r.Run(cctx, name, args...)
}

// resolveToken determines the GitHub token: the -token flag wins, then $GH_TOKEN, then
// $GITHUB_TOKEN, then `gh auth token`.
func resolveToken(ctx context.Context, r Runner, flagToken string) (string, error) {
	if flagToken != "" {
		return flagToken, nil
	}

	if t := os.Getenv("GH_TOKEN"); t != "" {
		return t, nil
	}

	if t := os.Getenv("GITHUB_TOKEN"); t != "" {
		return t, nil
	}

	out, err := runCmd(ctx, r, "gh", "auth", "token")
	if err != nil {
		return "", fmt.Errorf("resolve token: %w", err)
	}

	token := strings.TrimSpace(out)
	if token == "" {
		return "", errors.New("no GitHub token available (set -token, GH_TOKEN, GITHUB_TOKEN, or run `gh auth login`)")
	}

	return token, nil
}

// resolveRepo splits a repo into owner/name. When flagRepo is empty it falls back to
// `gh repo view`.
func resolveRepo(ctx context.Context, r Runner, flagRepo string) (owner, name string, err error) {
	spec := flagRepo
	if spec == "" {
		out, runErr := runCmd(ctx, r, "gh", "repo", "view", "--json", "nameWithOwner", "-q", ".nameWithOwner")
		if runErr != nil {
			return "", "", fmt.Errorf("resolve repo: %w", runErr)
		}

		spec = strings.TrimSpace(out)
	}

	owner, name, ok := strings.Cut(spec, "/")
	if !ok || owner == "" || name == "" {
		return "", "", fmt.Errorf("invalid repo %q: want owner/name", spec)
	}

	return owner, name, nil
}

// resolvePR returns the PR number: the -pr flag wins; otherwise it finds the PR for the
// current git branch via `git branch --show-current` and `gh pr list`.
func resolvePR(ctx context.Context, r Runner, flagPR int) (int, error) {
	if flagPR != 0 {
		return flagPR, nil
	}

	branchOut, err := runCmd(ctx, r, "git", "branch", "--show-current")
	if err != nil {
		return 0, fmt.Errorf("resolve pr: %w", err)
	}

	branch := strings.TrimSpace(branchOut)
	if branch == "" {
		return 0, errors.New("resolve pr: not on a branch (detached HEAD?); pass -pr")
	}

	out, err := runCmd(ctx, r, "gh", "pr", "list", "--head", branch, "--json", "number", "-q", ".[0].number")
	if err != nil {
		return 0, fmt.Errorf("resolve pr: %w", err)
	}

	s := strings.TrimSpace(out)
	if s == "" {
		return 0, fmt.Errorf("resolve pr: no open PR found for branch %q", branch)
	}

	n, err := strconv.Atoi(s)
	if err != nil {
		return 0, fmt.Errorf("resolve pr: parse %q: %w", s, err)
	}

	return n, nil
}
